File: apps/api/services/trust_battery.py
from enum import Enum
from datetime import datetime, timedelta
from dataclasses import dataclass, field
from typing import Optional


class TrustTier(str, Enum):
    PROBATION = "PROBATION"
    STANDARD = "STANDARD"
    CORE = "CORE"
    STRATEGIC = "STRATEGIC"


@dataclass
class TrustBattery:
    tier: TrustTier = TrustTier.PROBATION
    trust_score: int = 0
    consecutive_successes: int = 0
    consecutive_errors: int = 0
    days_in_current_tier: int = 0
    last_active_at: Optional[datetime] = field(default_factory=datetime.utcnow)

    def record_success(self) -> None:
        self.consecutive_successes += 1
        self.consecutive_errors = 0
        self.last_active_at = datetime.utcnow()

        if self.tier == TrustTier.PROBATION and self.consecutive_successes >= 3:
            self._try_upgrade()

    def record_error(self) -> None:
        self.consecutive_errors += 1
        self.consecutive_successes = 0

        if self.consecutive_errors >= 3:
            self._downgrade()

    def flag_fraud(self) -> None:
        self.tier = TrustTier.PROBATION
        self.trust_score = 0
        self.consecutive_successes = 0
        self.consecutive_errors = 0
        self.days_in_current_tier = 0

    def advance_days(self, days: int) -> None:
        self.days_in_current_tier += days
        self.last_active_at = datetime.utcnow()

        if self.tier == TrustTier.PROBATION and self.days_in_current_tier >= 30 and self.consecutive_successes >= 3:
            self._try_upgrade()

        if self.tier != TrustTier.PROBATION and (self.days_in_current_tier - days) >= 90:
            self._downgrade()

    def _try_upgrade(self) -> bool:
        upgrade_requirements = {
            TrustTier.PROBATION: (30, 3),
            TrustTier.STANDARD: (90, 10),
            TrustTier.CORE: (180, 0),
            TrustTier.STRATEGIC: (float('inf'), float('inf')),
        }

        days_required, tx_required = upgrade_requirements.get(self.tier, (float('inf'), float('inf')))

        if self.days_in_current_tier >= days_required:
            current_idx = list(TrustTier).index(self.tier)
            if current_idx < len(list(TrustTier)) - 1:
                self.tier = list(TrustTier)[current_idx + 1]
                self.days_in_current_tier = 0
                self.consecutive_successes = 0
                return True

        return False

    def _downgrade(self) -> None:
        if self.tier == TrustTier.PROBATION:
            return

        current_idx = list(TrustTier).index(self.tier)
        if current_idx > 0:
            self.tier = list(TrustTier)[current_idx - 1]
            self.consecutive_errors = 0
            self.days_in_current_tier = 0

    def should_downgrade(self) -> bool:
        if self.tier == TrustTier.PROBATION:
            return False

        if self.consecutive_errors >= 3:
            return True

        if self.last_active_at:
            days_inactive = (datetime.utcnow() - self.last_active_at).days
            if days_inactive >= 90:
                return True

        return False

    def to_dict(self) -> dict:
        return {
            "tier": self.tier.value,
            "trust_score": self.trust_score,
            "consecutive_successes": self.consecutive_successes,
            "consecutive_errors": self.consecutive_errors,
            "days_in_current_tier": self.days_in_current_tier,
            "last_active_at": self.last_active_at.isoformat() if self.last_active_at else None,
        }

    @classmethod
    def from_dict(cls, data: dict) -> "TrustBattery":
        return cls(
            tier=TrustTier(data.get("tier", "PROBATION")),
            trust_score=data.get("trust_score", 0),
            consecutive_successes=data.get("consecutive_successes", 0),
            consecutive_errors=data.get("consecutive_errors", 0),
            days_in_current_tier=data.get("days_in_current_tier", 0),
            last_active_at=datetime.fromisoformat(data["last_active_at"]) if data.get("last_active_at") else None,
        )

File: internal/domain/trust_battery.go
package domain

import (
	"time"
)

type TrustBattery struct {
	LastActiveAt         *time.Time `json:"last_active_at"`
	Tier                 TrustTier  `json:"tier"`
	TrustScore           int        `json:"trust_score"`
	ConsecutiveSuccesses int        `json:"consecutive_successes"`
	ConsecutiveErrors    int        `json:"consecutive_errors"`
	DaysInCurrentTier    int        `json:"days_in_current_tier"`
}

func NewTrustBattery() *TrustBattery {
	now := time.Now().UTC()
	return &TrustBattery{
		Tier:                 TrustTierProbation,
		TrustScore:           0,
		ConsecutiveSuccesses: 0,
		ConsecutiveErrors:    0,
		DaysInCurrentTier:    0,
		LastActiveAt:         &now,
	}
}

func (tb *TrustBattery) RecordSuccess() {
	tb.ConsecutiveSuccesses++
	tb.ConsecutiveErrors = 0
	now := time.Now().UTC()
	tb.LastActiveAt = &now

	if tb.Tier == TrustTierProbation && tb.ConsecutiveSuccesses >= 3 {
		tb.tryUpgrade()
	}
}

func (tb *TrustBattery) RecordError() {
	tb.ConsecutiveErrors++
	tb.ConsecutiveSuccesses = 0

	if tb.ConsecutiveErrors >= 3 {
		tb.downgrade()
	}
}

func (tb *TrustBattery) FraudFlagged() {
	// FraudFlagged moves to BLOCKED (terminal state)
	tb.Tier = TrustTierBlocked
	tb.TrustScore = 0
	tb.ConsecutiveSuccesses = 0
	tb.ConsecutiveErrors = 0
	tb.DaysInCurrentTier = 0
}

// AllChecksPass transitions from PROBATION to STANDARD
// Should be called when all verification checks pass
func (tb *TrustBattery) AllChecksPass() {
	if tb.Tier == TrustTierProbation {
		tb.Tier = TrustTierStandard
		tb.TrustScore = 31 // Minimum score for STANDARD tier
		tb.DaysInCurrentTier = 0
	}
}

// TransactionThresholdMet transitions based on transaction count:
// - 3 transactions: STANDARD → PREFERRED
// - 10 transactions: PREFERRED → STRATEGIC
func (tb *TrustBattery) TransactionThresholdMet(txCount int) {
	if tb.Tier == TrustTierStandard && txCount >= 3 {
		tb.Tier = TrustTierPreferred
		tb.TrustScore = 61 // Minimum score for PREFERRED tier
		tb.DaysInCurrentTier = 0
	} else if tb.Tier == TrustTierPreferred && txCount >= 10 {
		tb.Tier = TrustTierStrategic
		tb.TrustScore = 86 // Minimum score for STRATEGIC tier
		tb.DaysInCurrentTier = 0
	}
}

// DisputeFiled triggers downgrade:
// - PREFERRED → STANDARD
// - STRATEGIC → PREFERRED
func (tb *TrustBattery) DisputeFiled() {
	if tb.Tier == TrustTierPreferred {
		tb.Tier = TrustTierStandard
		tb.TrustScore = 31 // Downgrade to STANDARD score range
		tb.DaysInCurrentTier = 0
	} else if tb.Tier == TrustTierStrategic {
		tb.Tier = TrustTierPreferred
		tb.TrustScore = 61 // Downgrade to PREFERRED score range
		tb.DaysInCurrentTier = 0
	}
}

// InactivityDecay triggers downgrade after 180 days of inactivity:
// - STRATEGIC → PREFERRED
// - PREFERRED → STANDARD
func (tb *TrustBattery) InactivityDecay(days int) {
	if days >= 180 {
		if tb.Tier == TrustTierStrategic {
			tb.Tier = TrustTierPreferred
			tb.TrustScore = 61 // Downgrade to PREFERRED score range
			tb.DaysInCurrentTier = 0
		} else if tb.Tier == TrustTierPreferred {
			tb.Tier = TrustTierStandard
			tb.TrustScore = 31 // Downgrade to STANDARD score range
			tb.DaysInCurrentTier = 0
		}
	}
}

func (tb *TrustBattery) AdvanceDays(days int) {
	tb.DaysInCurrentTier += days
	now := time.Now().UTC()
	tb.LastActiveAt = &now

	if tb.Tier == TrustTierProbation && tb.DaysInCurrentTier >= 30 && tb.ConsecutiveSuccesses >= 3 {
		tb.tryUpgrade()
	}

	if tb.Tier != TrustTierProbation && (tb.DaysInCurrentTier-days) >= 90 {
		tb.downgrade()
	}
}

// ScoreBasedTier returns the trust tier based on trust score
func ScoreBasedTier(score int) TrustTier {
	switch {
	case score <= 30:
		return TrustTierProbation
	case score <= 60:
		return TrustTierStandard
	case score <= 85:
		return TrustTierPreferred
	default:
		return TrustTierStrategic
	}
}

func (tb *TrustBattery) tryUpgrade() bool {
	currentIdx := tb.tierIndex()
	tiers := []TrustTier{TrustTierProbation, TrustTierStandard, TrustTierPreferred, TrustTierStrategic}
	if currentIdx < len(tiers)-1 {
		tb.Tier = tiers[currentIdx+1]
		tb.DaysInCurrentTier = 0
		tb.ConsecutiveSuccesses = 0
		return true
	}
	return false
}

func (tb *TrustBattery) downgrade() {
	if tb.Tier == TrustTierProbation || tb.Tier == TrustTierBlocked {
		return
	}

	currentIdx := tb.tierIndex()
	tiers := []TrustTier{TrustTierProbation, TrustTierStandard, TrustTierPreferred, TrustTierStrategic}
	if currentIdx > 0 {
		tb.Tier = tiers[currentIdx-1]
		tb.ConsecutiveErrors = 0
		tb.DaysInCurrentTier = 0
	}
}

func (tb *TrustBattery) ShouldDowngrade() bool {
	if tb.Tier == TrustTierProbation {
		return false
	}

	if tb.ConsecutiveErrors >= 3 {
		return true
	}

	if tb.LastActiveAt != nil {
		daysInactive := int(time.Since(*tb.LastActiveAt).Hours() / 24)
		if daysInactive >= 90 {
			return true
		}
	}

	return false
}

func (tb *TrustBattery) tierIndex() int {
	tiers := []TrustTier{TrustTierProbation, TrustTierStandard, TrustTierPreferred, TrustTierStrategic, TrustTierBlocked}
	for i, t := range tiers {
		if t == tb.Tier {
			return i
		}
	}
	return 0
}

func (tb *TrustBattery) ToMap() map[string]interface{} {
	result := map[string]interface{}{
		"tier":                  string(tb.Tier),
		"trust_score":           tb.TrustScore,
		"consecutive_successes": tb.ConsecutiveSuccesses,
		"consecutive_errors":    tb.ConsecutiveErrors,
		"days_in_current_tier":  tb.DaysInCurrentTier,
	}

	if tb.LastActiveAt != nil {
		result["last_active_at"] = tb.LastActiveAt.Format(time.RFC3339)
	} else {
		result["last_active_at"] = nil
	}

	return result
}

func TrustBatteryFromMap(data map[string]interface{}) *TrustBattery {
	tb := &TrustBattery{}

	if tier, ok := data["tier"].(string); ok {
		tb.Tier = TrustTier(tier)
	} else {
		tb.Tier = TrustTierProbation
	}

	if v, ok := data["trust_score"].(float64); ok {
		tb.TrustScore = int(v)
	}
	if v, ok := data["consecutive_successes"].(float64); ok {
		tb.ConsecutiveSuccesses = int(v)
	}
	if v, ok := data["consecutive_errors"].(float64); ok {
		tb.ConsecutiveErrors = int(v)
	}
	if v, ok := data["days_in_current_tier"].(float64); ok {
		tb.DaysInCurrentTier = int(v)
	}

	if lastActive, ok := data["last_active_at"].(string); ok && lastActive != "" {
		if t, err := time.Parse(time.RFC3339, lastActive); err == nil {
			tb.LastActiveAt = &t
		}
	}

	return tb
}
